### PR TITLE
Allow migration from Gossip DNS to no DNS

### PR DESCRIPTION
**Pull request by hakman**

### CODE REVIEW

**Comment by johngmyers** [comment]
I think we might need to change https://github.com/kubernetes/kops/blob/85026145a1be32628815509cdeadcf7d8634dc1e/upup/pkg/fi/cloudup/openstack/cloud.go#L401

**Comment by hakman** [comment]
> I think we might need to change
> 
> https://github.com/kubernetes/kops/blob/85026145a1be32628815509cdeadcf7d8634dc1e/upup/pkg/fi/cloudup/openstack/cloud.go#L401

That part needs changing for sure, but `--dns=none` is not supported for OpenStack yet.

**Comment by johngmyers** [comment]
I'd prefer we don't leave behind a trap for later.

**Comment by hakman** [comment]
> I'd prefer we don't leave behind a trap for later.

Done in #14847

**Comment by hakman** [comment]
At the moment, the GCE implementation requires access from anywhere in the internet to work.
There is no way to allow only the server addresses as they are random public IPs.

File: upup/pkg/fi/cloudup/template_functions.go [hunk 1]
```diff
@@ -127,7 +127,7 @@ func (tf *TemplateFunctions) AddTo(dest template.FuncMap, secretStore fi.SecretS
 	dest["GossipDomains"] = func() []string {
 		var names []string
 
-		if cluster.IsGossip() {
+		if dns.IsGossipClusterName(cluster.Name) {
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/template_functions.go:130]: I don't understand this. If running NoneDNS in a cluster with a gossip name, `IsGossip()` is going to be false, yet all of the addons are going to enable the Gossip machinery?

Comment (hakman) [reply upup/pkg/fi/cloudup/template_functions.go:130]: This is not quite needed in general, only for the `[sig-auth] ServiceAccounts ServiceAccountIssuerDiscovery should support OIDC discovery of service account issuer [Conformance]` test to pass.

**Comment by johngmyers** [comment]
Could you explain what's going on here? It looks like a cluster with a Gossip name but enabling NoneDNS will, with this change, enable Gossip in both kops-controller and coredns. So it woudn't exactly be migrating from Gossip since it will still be doing some Gossip.

**Comment by hakman** [comment]
> Could you explain what's going on here? It looks like a cluster with a Gossip name but enabling NoneDNS will, with this change, enable Gossip in both kops-controller and coredns. So it woudn't exactly be migrating from Gossip since it will still be doing some Gossip.

Nothing, just some permissions to list endpoints and modify coredns configmap, which could be removed.
Most Gossip magic is done by Protokube, which is gone.

**Review by johngmyers — APPROVED** [review]

**Pull request merged**